### PR TITLE
Dependency fallbacks

### PR DESCRIPTION
**Pull request by xclaesse**
I had the crazy idea of using dict to create a list of dependency fallbacks:
```meson
cc = meson.get_compiler('c')
libfoo_fallbacks = [
  # Check for pkg-config name first
  'foo',
  # Check for an alternative pkg-config name
  'libfoo-1.0',
  # Check for a C library with headers
  {
    'compiler' : cc,
    'library' : 'foo',
    'headers' : ['foo.h']
  },
  # Check for an alternative library name
  {
    'compiler' : cc,
    'library' : 'libfoo-1.0',
    'headers' : ['foo.h']
  },
  # Fallback to a subproject
  {
    'subproject' : 'foo',
    'dependency' : 'libfoo_dep'
  },
]
libfoo_dep = dependecy(libfoo_fallbacks, required : get_option('some_feature'))
```

### CODE REVIEW

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -2969,21 +3003,59 @@ def _handle_featurenew_dependencies(self, name):
...
         if disabled:
             mlog.log('Dependency', mlog.bold(display_name), 'skipped: feature', mlog.bold(feature), 'disabled')
-            return DependencyHolder(NotFoundDependency(self.environment), self.subproject)
+            return self.notfound_dependency()
+
+        # Fallback only to subprojects in forcefallback mode, unless there is no
+        # subproject in the list.
+        if self.coredata.get_builtin_option('wrap_mode') == WrapMode.forcefallback:
+          subproject_fallbacks = [item for item in fallbacks if isinstance(item, dict) and 'subproject' in item]
+          if len(subproject_fallbacks) > 0:
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E111] indentation is not a multiple of four

File: mesonbuild/interpreter.py [hunk 2]
```diff
@@ -2969,21 +3003,59 @@ def _handle_featurenew_dependencies(self, name):
...
         disabled, required, feature = extract_required_kwarg(kwargs, self.subproject)
         if disabled:
             mlog.log('Dependency', mlog.bold(display_name), 'skipped: feature', mlog.bold(feature), 'disabled')
-            return DependencyHolder(NotFoundDependency(self.environment), self.subproject)
+            return self.notfound_dependency()
+
+        # Fallback only to subprojects in forcefallback mode, unless there is no
+        # subproject in the list.
+        if self.coredata.get_builtin_option('wrap_mode') == WrapMode.forcefallback:
+          subproject_fallbacks = [item for item in fallbacks if isinstance(item, dict) and 'subproject' in item]
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E111] indentation is not a multiple of four

File: mesonbuild/interpreter.py [hunk 3]
```diff
@@ -2969,21 +3003,59 @@ def _handle_featurenew_dependencies(self, name):
...
+          if len(subproject_fallbacks) > 0:
+              fallbacks = subproject_fallbacks
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E111] indentation is not a multiple of four

File: mesonbuild/interpreter.py [hunk 4]
```diff
@@ -3075,7 +3175,7 @@ def get_subproject_infos(self, kwargs):
 
     def dependency_fallback(self, name, kwargs):
         display_name = name if name else '(anonymous)'
-        if self.coredata.get_builtin_option('wrap_mode') in (WrapMode.nofallback, WrapMode.nodownload):
+        if self.coredata.get_builtin_option('wrap_mode')  == WrapMode.nofallback:
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E221] multiple spaces before operator

**Comment by xclaesse** [comment]
Judging from the examples @nirbheek posted in https://github.com/mesonbuild/meson/issues/2945#issuecomment-444961888 we could add a few extra helpers:
- The library fallback could take a list of libraries instead of just one, and they must all be found.
- find_library() would also have `has_function` keyword argument.

I think with those 2 small additions, this PR covers all the cases.

**Comment by xclaesse** [comment]
Yes, that's dictionaries with "magic" keys again, there are so useful :P

One way to tackle that is to add constructor for the dict:
```
png_fallbacks += cc.make_find_library_fallback('libpng16', has_headers : ['png.h'])
```

Or maybe add new object type that would basically be a python dict internally.

**Comment by xclaesse** [comment]
If we really want to rule out dictionaries with keys defined in the meson API, I would go with a new object type, something like that:
```meson
png_fallbacks = dependency_fallbacks('libpng16', 'libpng15', 'libpng14', 'libpng13', 'libpng12', 'libpng10')
foreach png: [ 'libpng16', 'libpng15', 'libpng14', 'libpng12', 'libpng13', 'libpng10' ]
  png_fallbacks.find_library(cc, png, has_headers : ['png.h'])
endforeach
png_fallbaks.subproject('png', 'png_dep')
png_dep = dependency(png_fallbacks, required : get_option('png'))
```

**Comment by xclaesse** [comment]
I created separated PRs for some of the commits:
https://github.com/mesonbuild/meson/pull/4658
https://github.com/mesonbuild/meson/pull/4659
https://github.com/mesonbuild/meson/pull/4660

**Comment by inigomartinez** [comment]
> I had the crazy idea of using dict to create a list of dependency fallbacks:
> 
> ```meson
> cc = meson.get_compiler('c')
> libfoo_fallbacks = [
>   # Check for pkg-config name first
>   'foo',
>   # Check for an alternative pkg-config name
>   'libfoo-1.0',
>   # Check for a C library with headers
>   {
>     'compiler' : cc,
>     'library' : 'foo',
>     'headers' : ['foo.h']
>   },
>   # Check for an alternative library name
>   {
>     'compiler' : cc,
>     'library' : 'libfoo-1.0',
>     'headers' : ['foo.h']
>   },
>   # Fallback to a subproject
>   {
>     'subproject' : 'foo',
>     'dependency' : 'libfoo_dep'
>   },
> ]
> libfoo_dep = dependecy(libfoo_fallbacks, required : get_option('some_feature'))
> ```

Could this use caso also be extended to `only headers`?

There are some cases where a function is available in `libc` but sometimes it needs a separate library. You can find this in [vinagre (also in some other code bases) with `socketpair`](https://gitlab.gnome.org/inigomartinez/vinagre/blob/meson/configure.ac#L178). [my approach](https://gitlab.gnome.org/GNOME/vinagre/merge_requests/4/diffs#0cc1139e3347f573ae1feee5b73dbc8a8a21fcfa_0_201) has been to create an empty array that holds any necessary library:

```
socket_deps = []
have_socket = cc.has_function('socketpair')
if not have_socket
  socket_deps += cc.find_library('socket', required: false)
  have_socket = cc.has_function('socketpair', dependencies: socket_deps)
endif
config_h.set('HAVE_SOCKETPAIR', have_socket)
```

I don't have a clear idea about how to fix this properly. It would be nice to have something similar to the `disabler` object that doesn't disable the creation of the targets where it's used but represents and empty/always not found dependency. If something like this existed, this could easily be solved and also `fallback` feature could be applied like this:

```
socket_fallbacks = [
  {
    `compiler`: cc,
    `dependency`: `null`, // null dependency which is always not found
    `function`: `socketpair`,
  },
  # check the existence of the function within a library
  {
    `compiler`: cc,
    `library`: `socket`
    `function`: `socketpair`,
  },
socket_dep = dependecy(socket_fallbacks, required : get_option('some_feature'))
```

**Comment by xclaesse** [comment]
On IRC @jpakkane suggested to define how a dependency needs to be found somehow in the native-file/cross-file that the user would provide, instead of coding that into build files.

I agree in some cases, for spacial platforms, some libraries could have a different name and only the user can know that. But I also think that the common cases the build should just work without the user knowing how each of the deps are provided. So I believe it's not one instead the other, but we actually need both solutions.

One other common issue is subprojects often repeat the lookup of a dependency, and each subproject could have a different set of fallbacks, resulting in some parts of the whole projecting linking on different libraries. One perfect example of this is libxml2 in GStreamer. Some parts look only for pkg-config file, other parts also look for lib+header, and some even fallback to wrapdb.  This can easily lead to building libxml2 as subproject because of a missing fallback to lib+header and another part link to the system library because it correctly found it. One other example of this is if you have 2 subprojects, the first depends on glib>=2.0 so it use the system one, then the 2nd needs glib>=2.58 and fallback to build it as subproject because system one is too old. You end  up using 2 different glib.

I think this all can be fixed with a new function: `override_dependency()` that would work just like `override_find_program()` because it's actually pretty much the same use-case.

When building glib as subproject, it will call `override_dependency('glib-2.0', glib_dep)` which will abort if anyone already called `dependency('glib-2.0')` and any subsequent call would return the subproject glib_dep instead of looking for pkg-config file on the system, even if subsequent calls does not have `fallback:['glib', 'glib_dep']`.

Similarly, the first time you call `dependency('libxml2')` with a set of fallbacks, any subsequent call will just return the same answer ignoring any other set of fallbacks. It will be the parent project's responsibility to make sure the first lookup is the one with the correct fallbacks.

And finally, any dependency declared in a cross/native file will be searched and cached first by meson, so any subsequent call will ignore fallbacks provided by build files.

File: mesonbuild/interpreter.py [hunk 5]
```diff
@@ -2065,6 +2091,121 @@ def get_cross_property_method(self, args, kwargs):
...
 
 
+class DependencyFallbacksHolder(InterpreterObject):
+    def __init__(self, interp):
+        InterpreterObject.__init__(self)
+        self.methods.update({'has_function': self.has_function_method,
+                             'find_library': self.find_library_method,
+                             'dependency': self.dependency_method,
+                             'subproject': self.subproject_method,
+                            })
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ closing bracket does not match visual indentation ([see on Sider](https://sider.review/gh/repos/19784232/pulls/4595#issue-4884433))

<details>

| Rule |
| - |
| E124 |

</details>
<sub>posted by Sider</sub>

File: mesonbuild/interpreter.py [hunk 6]
```diff
@@ -2065,6 +2091,121 @@ def get_cross_property_method(self, args, kwargs):
...
+        return None
+
+    def _do_find_library(self, name, display_name, kwargs, func_args, func_kwargs):
+        compiler = func_args[0]
+        return compiler.find_library_method(func_args[1:], func_kwargs)
+
+    def _do_dependency(self, name, display_name, kwargs, func_args, func_kwargs):
+        name = display_name = func_args[0]
+        fallbacks = DependencyFallbacksHolder(self.interpreter)
+        varname = func_kwargs.get('variable')
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ local variable 'varname' is assigned to but never used ([see on Sider](https://sider.review/gh/repos/19784232/pulls/4595#issue-4884434))

<details>

| Rule |
| - |
| F841 |

</details>
<sub>posted by Sider</sub>

**Comment by xclaesse** [comment]
I updated this idea, with new syntax:
```meson
fallbacks = dependency_fallbacks()
fallbacks.dependency('libxml2')
fallbacks.dependency('libxml-2.0')
fallbacks.find_library(cc, 'xml2', has_headers: 'libxml/tree.h')
fallbacks.find_library(cc, 'libxml2', has_headers: 'libxml/tree.h')
fallbacks.subproject('xml2', variable: 'xml2_dep')
d = dependency('xml2', fallbacks)
```

> ```
> socket_deps = []
> have_socket = cc.has_function('socketpair')
> if not have_socket
>   socket_deps += cc.find_library('socket', required: false)
>   have_socket = cc.has_function('socketpair', dependencies: socket_deps)
> endif
> config_h.set('HAVE_SOCKETPAIR', have_socket)
> ```

I'm not sure if you really need to check if the socket library has that function, but I guess it's a reasonable check. IMHO this could be simplified with this kind of syntax: `find_library('socket', has_function : 'socketpair')` in the same way we added `has_headers` kwargs. If we do that, then your snippet can be simplified to:
```meson
fallbacks = dependency_fallbacks()
fallbacks.has_function('socketpair')
fallbacks.find_library(cc, 'socket', has_function: 'socketpair', has_headers: 'sys/socket.h')
socket_deps = dependency('', fallbacks)
```

**Comment by xclaesse** [comment]
Note that this depends on https://github.com/mesonbuild/meson/pull/6203.

Open question: sometimes the caller wants to know which fallback it picked, to add some defines in their config.h. Not sure how to support that, or maybe just ignore that case?

**Comment by inigomartinez** [comment]
```
fallbacks = dependency_fallbacks()
fallbacks.dependency('libxml2')
fallbacks.dependency('libxml-2.0')
fallbacks.find_library(cc, 'xml2', has_headers: 'libxml/tree.h')
fallbacks.find_library(cc, 'libxml2', has_headers: 'libxml/tree.h')
fallbacks.subproject('xml2', variable: 'xml2_dep')
d = dependency('xml2', fallbacks)
```

What about using a combination of lists and dictionaries for this?

```
fallbacks = [
  {'dependency': 'libxml2'},
  {'dependency': 'libxml-2.0'},
  {'library': 'xml2', 'compiler': 'cc', 'has_headers': 'libxml/tree.h'},
  {'library': 'libxml2', 'compiler': 'cc', 'has_headers': 'libxml/tree.h'},
  {'subproject': 'xml2', 'variable': 'xml2_dep'},
]

d = dependency('xml2', fallbacks)
```

It might help reusing some of these _fallbacks_ if they are necessary multiple times.

**Comment by xclaesse** [comment]
That was my original proposal, but everyone hated it.

**Comment by inigomartinez** [comment]
You are right. I have forgotten to review your original proposal and I overlooked the change from dict keys to method calls.

Actually I didn't hate your original proposal.

**Comment by scivision** [comment]
this seems better than the current things I've been doing:
* adding each custom dependency inside Meson itself
* nested foreach, if
It's even more than I'd hoped for : `dependency('foo', 'bar', ...)`

**Comment by dcbaker** [comment]
I think sometimes the custom dependencies inside meson make sense, from the point of view of our "don't make everyone re-implement the wheel, wrong" philosophy.

I'm not entirely convinced that this syntax is the great, but  the current approach of deeply nested `if dep.found() else` is much worse.

What about something like this (maybe you've thought about this and thought of corners I haven't:

```meson
dependency('zlib', 'ZLIB', 'z', libraries  : ['z'], has_headers : ['zlib.h'], compiler : cc, fallback : ['zlib', 'dep_zlib'])
```

Where you can pass an unlimited number of strings for dependency to try (which would also be nice to have because cmake and pkg-config often disagree about how to name something). and then a libraries and has_headers.

It's a little DRYer than your approach I think, and the ability to specify multiple dependency names to try would be useful in and of itself.

**Comment by xclaesse** [comment]
@dcbaker I don't like the syntax I proposed neither, tbh. I feel this is a complex stuff to do, and there are no perfect solution. tbh, your proposal could be good enough too, what I find a bit weird is it's not obvious that `libraries:` kwarg is a fallback and it's not obvious that if you pass `libraries:` you also have to pass `compiler:`. But that maybe that's just nitpicking. You propose plurial "libraries", is it a list of fallbacks, or all of them must be found?

While we are at proposing alternatives, I also thought about:
```meson
dependency('zlib', 'ZLIB', 'z',
  cc.fallback_find_library('z', has_headers : ['zlib.h']),
  cc.fallback_find_library('zlib1', has_headers : ['zlib.h']),
  fallback : ['zlib', 'dep_zlib']
)
```
Basically similar to my previous proposal, but with methods directly on the compiler object instead of on an intermediary fallback object.

**Comment by mensinda** [comment]
From all the syntax proposals, I like [this one](https://github.com/mesonbuild/meson/pull/4595#issuecomment-559836031) best. However, I don't like having to repeat the variable name. So, since everyone is proposing a syntax, here is what I came up with:

```meson
dep = dependency(
  dependency_fallbacks()
    .dependency('zlib')
    .dependency('ZLIB')
    .dependency('z')
    .find_library(cc, 'z',     has_headers: ['zlib.h'])
    .find_library(cc, 'zlib1', has_headers: ['zlib.h'])
    .subproject('zlib', variable: 'zlib_dep'),
  version: '>=1.2.3'
)
```

The basic idea is that each method of `dependency_fallbacks()` returns itself, thus allowing chaining the fallback methods. This might require some slight tweaks in the parser/interpreter to handle the newlines between the object and method.

We could go even one step further and have `dependency()` without any arguments return a temporary object, similar to `dependency_fallbacks()`. The fallbacks could be added the same way as in the `dependency_fallbacks()` solution. Finally, a `find()` method has to be called, which will return the real dependency object. After calling `find()` the method, the temporary fallbacks dependency object would become immutable, and using it in any way would result in an error.

This proposal would look something like this:

```
dep = dependency()
  .dependency('zlib')
  .dependency('ZLIB')
  .dependency('z')
  .find_library(cc, 'z',     has_headers: ['zlib.h'])
  .find_library(cc, 'zlib1', has_headers: ['zlib.h'])
  .subproject('zlib', variable: 'zlib_dep')
  .find(version: '>=1.2.3', disabler: true)  # accepts all kwargs from dependency()
```

Regardless of what solution is chosen, I don't like adding new kwargs to `dependency()` for this. IMHO kwargs should be mostly independent and not rely on each other (for instance, specifying `libraries` would be an error without also providing `compiler`).

**Comment by xclaesse** [comment]
I personally like 'builder' pattern, but it has never been used in Meson, dunno how @jpakkane would feel about it (your 2nd example could be a bit extreme). otoh, what I like with new methods like `fallback_find_library()` is we can easily put them into a list, which is convenient to append items, etc.

**Comment by mensinda** [comment]
I don't really get your point with the benefits of lists. Appending fallbacks with the builder pattern is still possible since one "benefit" of it is that you are not required to use it. So this would also work:

```meson
df = dependency_fallbacks()
  .dependency('zlib')
  .dependency('ZLIB')
  .dependency('z')

foreach i : ['z', 'zlib1']
  df.find_library(cc, i, has_headers: ['zlib.h'])
endforeach

dep = dependency(df)
```

**Comment by xclaesse** [comment]
> I don't really get your point with the benefits of lists. Appending fallbacks with the builder pattern is still possible since one "benefit" of it is that you are not required to use it.

It's pretty much the same indeed. I just feel like using native lists from the language is more natural, but it doesn't really matter.

**Comment by jpakkane** [comment]
What I don't like about these is that they get all too complicated to read and somewhat contrived. Maybe it would be better to keep the two common cases (only use found dependencies and building your own) separate. Mixing the two can be a bit more to type if the common case becomes easier.

Getting dependencies with many names should really be just `d = dependency('zlib', 'ZLIB', 'bob', ...)`, mirroring how `find_program` works.

We should have a better way of building dependency objects for things found manually. Something like `d = cc.manual_dependency(libraries: 'z', headers: 'zlib.h')`.

AFAICT there are not that many dependencies that need to be found with both pkg-config and by manually looking up the lib + headers.

**Comment by xclaesse** [comment]
> Getting dependencies with many names should really be just `d = dependency('zlib', 'ZLIB', 'bob', ...)`, mirroring how `find_program` works.

Agreed, my proposal in https://github.com/mesonbuild/meson/pull/4595#issuecomment-571739966 works like that. I think it's a good start even if we do only that part.

> We should have a better way of building dependency objects for things found manually. Something like `d = cc.manual_dependency(libraries: 'z', headers: 'zlib.h')`.

Not sure what you mean here, what's the difference with already working `d = cc.find_library('z', has_headers: 'zlib.h')` ? 

> AFAICT there are not that many dependencies that need to be found with both pkg-config and by manually looking up the lib + headers.

Sadly, reading what many projects does (many are linked in conversation above), there are actually many of them... The problem is that in practice pkg-config is only really available only on Linux distro. Platform SDKs rarely ship pc files (e.g. Android).

**Comment by mensinda** [comment]
There doesn't appear to be a good solution that both keeps the simple use case simple and also allows complex stuff like `find_library`. (At least I can't find one). So, why don't we do both? Make a simple `dependency()` with multiple names **and** a `dependency_fallbacks()` (or list-based) solution.

I am aware that this introduces two new APIs to do basically the same thing (which is also not ideal). However, we could avoid some confusion if we make them mutually exclusive. For instance, you shouldn't be able to specify any dependency names if you also have a `dependency_fallbacks()`.

```meson
df = dependency_fallbacks()
  .dependency('zlib')
  .dependency('z')
  .find_library(cc, 'z',     has_headers: ['zlib.h'])
  .find_library(cc, 'zlib1', has_headers: ['zlib.h'])

dep0 = dependency('zlib', 'ZLIB', 'z',  version: '>=1.2.3')   # OK
dep1 = dependency(df,  version: '>=1.2.3')                    # OK
dep3 = dependency('ZLIB', df,  version: '>=1.2.3')            # ERROR
```

> Sadly, reading what many projects does (many are linked in conversation above), there are actually many of them... The problem is that in practice pkg-config is only really available only on Linux distro. Platform SDKs rarely ship pc files (e.g. Android).

I also agree with this point. So, just having `dependency('zlib', 'ZLIB', 'z')` would still be an improvement, however, we might end up with multiple ways of doing things if we also want to support `find_library` (again I am not 100% sure if this is a bad thing in this case if done right).

**Comment by ePirat** [comment]
Another maybe awful idea: If you make dependency and the like lazily evaluated you could do:

```meson
fallbacks = [
  dependency('libxml2'),
  dependency('libxml-2.0'),
  cc.find_library('xml2', has_headers: 'libxml/tree.h'),
  cc.find_library('libxml2', has_headers: 'libxml/tree.h'),
  subproject('xml2', variable: 'xml2_dep')
]
d = dependency('xml2', fallbacks: fallbacks)
```

**Comment by bonzini** [comment]
Looking at the problem from another angle, one language feature that I find missing is a less repetitive version of

```
...
if not f.found()
  f = ...
endif
if not f.found()
  f = ...
endif
```

Perhaps something like

```
do
  ...
then if not f.found()
  ...
then if not f.found()
  ...
enddo
```

Here, `then if not f.found()` would be kind of a break statement&mdash;a shorter and more readable version of `if f.found() break endif`&mdash;but note that I am _not_ proposing (at least not in the first implementation) to allow `break` inside this new `do` statement. You could then write:

```
do
  d = dependency('libxml2')
then if not d.found()
  d = dependency('libxml-2.0')
then if not d.found()
  d = cc.find_library('xml2', has_headers: 'libxml/tree.h')
then if not d.found()
  d = cc.find_library('libxml2', has_headers: 'libxml/tree.h')
then if not d.found()
  d = dependency('libxml2', fallback: ['xml2', 'xml2_dep'])
enddo
```

This new language construct could also be useful with positive conditions, in case one test needs to find multiple libraries:

```
mydep = dependency('', required: false)
do
  lib1 = cc.find_library(...)
then if lib1.found()
  lib2 = cc.find_library(...)
then if lib2.found()
  mydep = declare_dependency(link_with: lib1, lib2)
enddo
```

Another possibility could be to add a string argument to "do", like

```
do 'Searching for libxml2'
  ...
enddo
```

that would be printed with indentation:

```
Searching for libxml2:
  Dependency libxml2: NO
  Dependency libxml2-2.0: NO
  Finding library libxml2: YES
```

**Comment by xclaesse** [comment]
@bonzini One thing I want with fallbacks is to cache the result for consistency.
```
# Assume ZLIB is found but not zlib.
zlib_dep = dependency('zlib', 'ZLIB')
...
# In another subproject...
# Return the ZLIB dependency because it was cached earlier.
dependency('zlib')
```

I personally think we should go with `df = dependency_fallbacks()` because it is the most flexible option. We can add as many methods and kwargs as we want with this new type of object. That should be future proof.

**Comment by xclaesse** [comment]
Let's close this, I created a new PR with a variant of the dependency_fallback() API discussed here.

https://github.com/mesonbuild/meson/pull/8699
